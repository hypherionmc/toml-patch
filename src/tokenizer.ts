export enum TokenType {
  Bracket = 'Bracket',
  Curly = 'Curly',
  Equal = 'Equal',
  Comma = 'Comma',
  Comment = 'Comment',
  String = 'String'
}

export interface Token {
  token_type: TokenType;
  raw: string;
  loc: Location;
}

export interface Location {
  start: Position;
  end: Position;
}

export interface Position {
  // Note: line is 1-indexed while column is 0-indexed
  line: number;
  column: number;
}

const IS_WHITESPACE = /\s/;
const IS_NEW_LINE = /(\r\n|\n)/;
const IS_VALID_LEADING_CHARACTER = /[\w,\d,\",\',\+,\-,\_]/;
const DOUBLE_QUOTE = `"`;
const SINGLE_QUOTE = `'`;
const SPACE = ' ';
const ESCAPE = '\\';
const IS_FULL_DATE = /(\d{4})-(\d{2})-(\d{2})/;
const IS_FULL_TIME = /(\d{2}):(\d{2}):(\d{2})/;

export function tokenize(input: string): Token[] {
  let current = 0;
  const tokens: Token[] = [];

  const lines = findLines(input);
  const location = (start: number, end: number) => {
    return {
      start: findPosition(lines, start),
      end: findPosition(lines, end)
    };
  };

  while (current < input.length) {
    let char = input[current];

    const next = (step: number = 1) => {
      current += step;
      char = input[current];
    };
    const special = (token_type: TokenType) => {
      tokens.push({ token_type, raw: char, loc: location(current, current + 1) });
      current++;
    };

    if (IS_WHITESPACE.test(char)) {
      current++;
      continue;
    }

    // Handle special characters: [, ], {, }, =, comma
    if (char === '[' || char === ']') {
      special(TokenType.Bracket);
      continue;
    }
    if (char === '{' || char === '}') {
      special(TokenType.Curly);
      continue;
    }
    if (char === '=') {
      special(TokenType.Equal);
      continue;
    }
    if (char === ',') {
      special(TokenType.Comma);
      continue;
    }

    // Handle comments = # -> EOL
    if (char === '#') {
      const start = current;
      let raw = '';
      while (!IS_NEW_LINE.test(char) && current < input.length) {
        raw += char;
        next();
      }

      tokens.push({
        token_type: TokenType.Comment,
        raw,
        loc: location(start, current)
      });
      continue;
    }

    // Multi-line literals or strings = no escaping
    const multiline_char =
      checkThree(input, current, SINGLE_QUOTE) || checkThree(input, current, DOUBLE_QUOTE);
    if (multiline_char) {
      const start = current;
      let raw = multiline_char + multiline_char + multiline_char;
      next(3);

      while (!checkThree(input, current, multiline_char) && current < input.length) {
        raw += char;
        next();
      }

      raw += multiline_char + multiline_char + multiline_char;
      current += 3;

      tokens.push({
        token_type: TokenType.String,
        raw,
        loc: location(start, current)
      });

      current++;
      continue;
    }

    // Remaining possibilities: keys, strings, literals, integer, float, boolean
    //
    // Special cases:
    // "..." -> quoted
    // '...' -> quoted
    // "...".'...' -> bare
    // 0000-00-00 00:00:00 -> bare
    //
    // See https://github.com/toml-lang/toml#offset-date-time
    //
    // | For the sake of readability, you may replace the T delimiter between date and time with a space (as permitted by RFC 3339 section 5.6).
    // | `odt4 = 1979-05-27 07:32:00Z`
    //
    // From RFC 3339:
    //
    // | NOTE: ISO 8601 defines date and time separated by "T".
    // | Applications using this syntax may choose, for the sake of
    // | readability, to specify a full-date and full-time separated by
    // | (say) a space character.

    // First, check for invalid characters
    if (!IS_VALID_LEADING_CHARACTER.test(char)) {
      throw new Error(`Unsupported character "${char}" found at ${current}`);
    }

    const start = current;
    let raw = '';
    let double_quoted = false;
    let single_quoted = false;

    while (current < input.length) {
      if (char === DOUBLE_QUOTE) double_quoted = !double_quoted;
      if (char === SINGLE_QUOTE) single_quoted = !single_quoted;

      raw += char;
      next();

      // If next character is escape and currently double-quoted,
      // check for escaped quote
      if (double_quoted && char === ESCAPE) {
        if (input[current + 1] === DOUBLE_QUOTE) {
          raw += ESCAPE + DOUBLE_QUOTE;
          next(2);
        }
      }

      // If next character is IS_WHITESPACE,
      // check if raw is full date and following is full time
      if (char === SPACE && IS_FULL_DATE.test(raw)) {
        const possibly_time = input.substr(current + 1, 8);
        if (IS_FULL_TIME.test(possibly_time)) {
          raw += SPACE;
          next();
        }
      }

      if (
        !(double_quoted || single_quoted) &&
        (IS_WHITESPACE.test(char) || char === ',' || char === ']')
      )
        break;
    }

    if (double_quoted) {
      throw new Error(`Un-closed string found starting at ${start} (${raw})`);
    }
    if (single_quoted) {
      throw new Error(`Un-closed string literal found starting at ${start} (${raw})`);
    }

    tokens.push({
      token_type: TokenType.String,
      raw,
      loc: location(start, current)
    });
  }

  return tokens;
}

function checkThree(input: string, current: number, check: string): false | string {
  return (
    input[current] === check &&
    input[current + 1] === check &&
    input[current + 2] === check &&
    check
  );
}

export function findLines(input: string): number[] {
  // exec is stateful, so create new regexp each time
  const BY_NEW_LINE = /[\r\n|\n]/g;
  const indexes: number[] = [];

  let match;
  while ((match = BY_NEW_LINE.exec(input)) != null) {
    indexes.push(match.index);
  }

  return indexes.concat([input.length + 1]);
}

// abc\ndef\ng
// 0123 4567 8
//      012
//           0
//
// lines = [3, 7, 9]
//
// c = 2: 0 -> 1, 2 - (undefined + 1 || 0) = 2
//     3: 0 -> 1, 3 - (undefined + 1 || 0) = 3
// e = 5: 1 -> 2, 5 - (3 + 1 || 0) = 1
// g = 8: 2 -> 3, 8 - (7 + 1 || 0) = 0

export function findPosition(lines: number[], index: number): Position {
  const line = lines.findIndex(line_index => line_index >= index) + 1;
  const column = index - (lines[line - 2] + 1 || 0);

  return { line, column };
}
