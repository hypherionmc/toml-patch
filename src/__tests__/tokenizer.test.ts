import { tokenize } from '../tokenizer';

test('should tokenize simple input', () => {
  expect(tokenize(`a = "b"`)).toMatchSnapshot();
});

test('should tokenize multiline strings', () => {
  expect(
    tokenize(`
    a = "b"
    c = 'd'
    e = """
      f
    """
    g = '''
      h
    '''
    "i".'j'.k = "l"
  `)
  ).toMatchSnapshot();
});

const complex = `# This is a TOML document.

title = "TOML Example"

[owner]
name = "Tom Preston-Werner"
dob = 1979-05-27T07:32:00-08:00 # First class dates

[database]
server = "192.168.1.1"
ports = [ 8001, 8001, 8002 ]
connection_max = 5000
enabled = true

[servers]

  # Indentation (tabs and/or spaces) is allowed but not required
  [servers.alpha]
  ip = "10.0.0.1"
  dc = "eqdc10"

  [servers.beta]
  ip = "10.0.0.2"
  dc = "eqdc10"

[clients]
data = [ ["gamma", "delta"], [1, 2] ]

# Line breaks are OK when inside arrays
hosts = [
  "alpha",
  "omega"
]

a.'b'."c" = "d"
date_space = 2018-01-01 08:00:00`;

test('should tokenize complex input', () => {
  expect(tokenize(complex)).toMatchSnapshot();
});
