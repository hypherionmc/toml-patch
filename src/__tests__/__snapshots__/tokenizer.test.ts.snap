// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should tokenize complex input 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 26,
        "line": 1,
      },
      "start": Object {
        "column": 0,
        "line": 1,
      },
    },
    "raw": "# This is a TOML document.",
    "token_type": "Comment",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 3,
      },
      "start": Object {
        "column": 0,
        "line": 3,
      },
    },
    "raw": "title",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 3,
      },
      "start": Object {
        "column": 6,
        "line": 3,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 22,
        "line": 3,
      },
      "start": Object {
        "column": 8,
        "line": 3,
      },
    },
    "raw": "\\"TOML Example\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 5,
      },
      "start": Object {
        "column": 0,
        "line": 5,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 5,
      },
      "start": Object {
        "column": 1,
        "line": 5,
      },
    },
    "raw": "owner",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 5,
      },
      "start": Object {
        "column": 6,
        "line": 5,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 6,
      },
      "start": Object {
        "column": 0,
        "line": 6,
      },
    },
    "raw": "name",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 6,
      },
      "start": Object {
        "column": 5,
        "line": 6,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 27,
        "line": 6,
      },
      "start": Object {
        "column": 7,
        "line": 6,
      },
    },
    "raw": "\\"Tom Preston-Werner\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 3,
        "line": 7,
      },
      "start": Object {
        "column": 0,
        "line": 7,
      },
    },
    "raw": "dob",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 7,
      },
      "start": Object {
        "column": 4,
        "line": 7,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 31,
        "line": 7,
      },
      "start": Object {
        "column": 6,
        "line": 7,
      },
    },
    "raw": "1979-05-27T07:32:00-08:00",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 51,
        "line": 7,
      },
      "start": Object {
        "column": 32,
        "line": 7,
      },
    },
    "raw": "# First class dates",
    "token_type": "Comment",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 9,
      },
      "start": Object {
        "column": 0,
        "line": 9,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 9,
      },
      "start": Object {
        "column": 1,
        "line": 9,
      },
    },
    "raw": "database",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 10,
        "line": 9,
      },
      "start": Object {
        "column": 9,
        "line": 9,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 10,
      },
      "start": Object {
        "column": 0,
        "line": 10,
      },
    },
    "raw": "server",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 8,
        "line": 10,
      },
      "start": Object {
        "column": 7,
        "line": 10,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 22,
        "line": 10,
      },
      "start": Object {
        "column": 9,
        "line": 10,
      },
    },
    "raw": "\\"192.168.1.1\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 11,
      },
      "start": Object {
        "column": 0,
        "line": 11,
      },
    },
    "raw": "ports",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 11,
      },
      "start": Object {
        "column": 6,
        "line": 11,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 11,
      },
      "start": Object {
        "column": 8,
        "line": 11,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 14,
        "line": 11,
      },
      "start": Object {
        "column": 10,
        "line": 11,
      },
    },
    "raw": "8001",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 11,
      },
      "start": Object {
        "column": 14,
        "line": 11,
      },
    },
    "raw": ",",
    "token_type": "Comma",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 20,
        "line": 11,
      },
      "start": Object {
        "column": 16,
        "line": 11,
      },
    },
    "raw": "8001",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 21,
        "line": 11,
      },
      "start": Object {
        "column": 20,
        "line": 11,
      },
    },
    "raw": ",",
    "token_type": "Comma",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 26,
        "line": 11,
      },
      "start": Object {
        "column": 22,
        "line": 11,
      },
    },
    "raw": "8002",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 28,
        "line": 11,
      },
      "start": Object {
        "column": 27,
        "line": 11,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 14,
        "line": 12,
      },
      "start": Object {
        "column": 0,
        "line": 12,
      },
    },
    "raw": "connection_max",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 16,
        "line": 12,
      },
      "start": Object {
        "column": 15,
        "line": 12,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 21,
        "line": 12,
      },
      "start": Object {
        "column": 17,
        "line": 12,
      },
    },
    "raw": "5000",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 13,
      },
      "start": Object {
        "column": 0,
        "line": 13,
      },
    },
    "raw": "enabled",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 13,
      },
      "start": Object {
        "column": 8,
        "line": 13,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 14,
        "line": 13,
      },
      "start": Object {
        "column": 10,
        "line": 13,
      },
    },
    "raw": "true",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 15,
      },
      "start": Object {
        "column": 0,
        "line": 15,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 8,
        "line": 15,
      },
      "start": Object {
        "column": 1,
        "line": 15,
      },
    },
    "raw": "servers",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 15,
      },
      "start": Object {
        "column": 8,
        "line": 15,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 64,
        "line": 17,
      },
      "start": Object {
        "column": 2,
        "line": 17,
      },
    },
    "raw": "# Indentation (tabs and/or spaces) is allowed but not required",
    "token_type": "Comment",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 3,
        "line": 18,
      },
      "start": Object {
        "column": 2,
        "line": 18,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 16,
        "line": 18,
      },
      "start": Object {
        "column": 3,
        "line": 18,
      },
    },
    "raw": "servers.alpha",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 17,
        "line": 18,
      },
      "start": Object {
        "column": 16,
        "line": 18,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 19,
      },
      "start": Object {
        "column": 2,
        "line": 19,
      },
    },
    "raw": "ip",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 19,
      },
      "start": Object {
        "column": 5,
        "line": 19,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 17,
        "line": 19,
      },
      "start": Object {
        "column": 7,
        "line": 19,
      },
    },
    "raw": "\\"10.0.0.1\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 20,
      },
      "start": Object {
        "column": 2,
        "line": 20,
      },
    },
    "raw": "dc",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 20,
      },
      "start": Object {
        "column": 5,
        "line": 20,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 20,
      },
      "start": Object {
        "column": 7,
        "line": 20,
      },
    },
    "raw": "\\"eqdc10\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 3,
        "line": 22,
      },
      "start": Object {
        "column": 2,
        "line": 22,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 22,
      },
      "start": Object {
        "column": 3,
        "line": 22,
      },
    },
    "raw": "servers.beta",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 16,
        "line": 22,
      },
      "start": Object {
        "column": 15,
        "line": 22,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 23,
      },
      "start": Object {
        "column": 2,
        "line": 23,
      },
    },
    "raw": "ip",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 23,
      },
      "start": Object {
        "column": 5,
        "line": 23,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 17,
        "line": 23,
      },
      "start": Object {
        "column": 7,
        "line": 23,
      },
    },
    "raw": "\\"10.0.0.2\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 24,
      },
      "start": Object {
        "column": 2,
        "line": 24,
      },
    },
    "raw": "dc",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 24,
      },
      "start": Object {
        "column": 5,
        "line": 24,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 24,
      },
      "start": Object {
        "column": 7,
        "line": 24,
      },
    },
    "raw": "\\"eqdc10\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 26,
      },
      "start": Object {
        "column": 0,
        "line": 26,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 8,
        "line": 26,
      },
      "start": Object {
        "column": 1,
        "line": 26,
      },
    },
    "raw": "clients",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 26,
      },
      "start": Object {
        "column": 8,
        "line": 26,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 27,
      },
      "start": Object {
        "column": 0,
        "line": 27,
      },
    },
    "raw": "data",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 27,
      },
      "start": Object {
        "column": 5,
        "line": 27,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 8,
        "line": 27,
      },
      "start": Object {
        "column": 7,
        "line": 27,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 10,
        "line": 27,
      },
      "start": Object {
        "column": 9,
        "line": 27,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 17,
        "line": 27,
      },
      "start": Object {
        "column": 10,
        "line": 27,
      },
    },
    "raw": "\\"gamma\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 18,
        "line": 27,
      },
      "start": Object {
        "column": 17,
        "line": 27,
      },
    },
    "raw": ",",
    "token_type": "Comma",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 26,
        "line": 27,
      },
      "start": Object {
        "column": 19,
        "line": 27,
      },
    },
    "raw": "\\"delta\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 27,
        "line": 27,
      },
      "start": Object {
        "column": 26,
        "line": 27,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 28,
        "line": 27,
      },
      "start": Object {
        "column": 27,
        "line": 27,
      },
    },
    "raw": ",",
    "token_type": "Comma",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 30,
        "line": 27,
      },
      "start": Object {
        "column": 29,
        "line": 27,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 31,
        "line": 27,
      },
      "start": Object {
        "column": 30,
        "line": 27,
      },
    },
    "raw": "1",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 32,
        "line": 27,
      },
      "start": Object {
        "column": 31,
        "line": 27,
      },
    },
    "raw": ",",
    "token_type": "Comma",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 34,
        "line": 27,
      },
      "start": Object {
        "column": 33,
        "line": 27,
      },
    },
    "raw": "2",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 35,
        "line": 27,
      },
      "start": Object {
        "column": 34,
        "line": 27,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 37,
        "line": 27,
      },
      "start": Object {
        "column": 36,
        "line": 27,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 39,
        "line": 29,
      },
      "start": Object {
        "column": 0,
        "line": 29,
      },
    },
    "raw": "# Line breaks are OK when inside arrays",
    "token_type": "Comment",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 30,
      },
      "start": Object {
        "column": 0,
        "line": 30,
      },
    },
    "raw": "hosts",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 30,
      },
      "start": Object {
        "column": 6,
        "line": 30,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 30,
      },
      "start": Object {
        "column": 8,
        "line": 30,
      },
    },
    "raw": "[",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 31,
      },
      "start": Object {
        "column": 2,
        "line": 31,
      },
    },
    "raw": "\\"alpha\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 10,
        "line": 31,
      },
      "start": Object {
        "column": 9,
        "line": 31,
      },
    },
    "raw": ",",
    "token_type": "Comma",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 32,
      },
      "start": Object {
        "column": 2,
        "line": 32,
      },
    },
    "raw": "\\"omega\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 33,
      },
      "start": Object {
        "column": 0,
        "line": 33,
      },
    },
    "raw": "]",
    "token_type": "Bracket",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 35,
      },
      "start": Object {
        "column": 0,
        "line": 35,
      },
    },
    "raw": "a.'b'.\\"c\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 11,
        "line": 35,
      },
      "start": Object {
        "column": 10,
        "line": 35,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 35,
      },
      "start": Object {
        "column": 12,
        "line": 35,
      },
    },
    "raw": "\\"d\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 10,
        "line": 36,
      },
      "start": Object {
        "column": 0,
        "line": 36,
      },
    },
    "raw": "date_space",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 36,
      },
      "start": Object {
        "column": 11,
        "line": 36,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 32,
        "line": 36,
      },
      "start": Object {
        "column": 13,
        "line": 36,
      },
    },
    "raw": "2018-01-01 08:00:00",
    "token_type": "String",
  },
]
`;

exports[`should tokenize multiline strings 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 2,
      },
      "start": Object {
        "column": 4,
        "line": 2,
      },
    },
    "raw": "a",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 2,
      },
      "start": Object {
        "column": 6,
        "line": 2,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 11,
        "line": 2,
      },
      "start": Object {
        "column": 8,
        "line": 2,
      },
    },
    "raw": "\\"b\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 3,
      },
      "start": Object {
        "column": 4,
        "line": 3,
      },
    },
    "raw": "c",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 3,
      },
      "start": Object {
        "column": 6,
        "line": 3,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 11,
        "line": 3,
      },
      "start": Object {
        "column": 8,
        "line": 3,
      },
    },
    "raw": "'d'",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 4,
      },
      "start": Object {
        "column": 4,
        "line": 4,
      },
    },
    "raw": "e",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 4,
      },
      "start": Object {
        "column": 6,
        "line": 4,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 6,
      },
      "start": Object {
        "column": 8,
        "line": 4,
      },
    },
    "raw": "\\"\\"\\"
      f
    \\"\\"\\"",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 7,
      },
      "start": Object {
        "column": 4,
        "line": 7,
      },
    },
    "raw": "g",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 7,
      },
      "start": Object {
        "column": 6,
        "line": 7,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 9,
      },
      "start": Object {
        "column": 8,
        "line": 7,
      },
    },
    "raw": "'''
      h
    '''",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 13,
        "line": 10,
      },
      "start": Object {
        "column": 4,
        "line": 10,
      },
    },
    "raw": "\\"i\\".'j'.k",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 10,
      },
      "start": Object {
        "column": 14,
        "line": 10,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 19,
        "line": 10,
      },
      "start": Object {
        "column": 16,
        "line": 10,
      },
    },
    "raw": "\\"l\\"",
    "token_type": "String",
  },
]
`;

exports[`should tokenize simple input 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 1,
      },
      "start": Object {
        "column": 0,
        "line": 1,
      },
    },
    "raw": "a",
    "token_type": "String",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 3,
        "line": 1,
      },
      "start": Object {
        "column": 2,
        "line": 1,
      },
    },
    "raw": "=",
    "token_type": "Equal",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 1,
      },
      "start": Object {
        "column": 4,
        "line": 1,
      },
    },
    "raw": "\\"b\\"",
    "token_type": "String",
  },
]
`;
